src/cachingreaderchunk.cpp

```diff
@@ -0,0 +1,186 @@
+            // faulty files: Try to skip some samples up to the requested
+            // seek position. But only skip twice as many frames/samples
+            // as have been requested to avoid decoding great portions of
+            // the file for small read requests on seek errors.
+            const SINT framesToSkip = frameIndex - seekFrameIndex;
+            if (framesToSkip <= (2 * framesToRead)) {
+                seekFrameIndex += pAudioSource->skipSampleFrames(framesToSkip);
+            }
+        }
+        if (frameIndex != seekFrameIndex) {
+            // Unexpected/premature end of file -> prevent further
+            // seeks beyond the current seek position
+            *pMaxReadableFrameIndex = math_min(seekFrameIndex, *pMaxReadableFrameIndex);
+            // Don't read any samples on a seek failure!
+            m_frameCount = 0;
+            return m_frameCount;
+        }
+    }
+
+    DEBUG_ASSERT(frameIndex == seekFrameIndex);
+    DEBUG_ASSERT(CachingReaderChunk::kChannels
+            == Mixxx::AudioSource::kChannelCountStereo);
+    m_frameCount = pAudioSource->readSampleFramesStereo(
+            framesToRead, m_sampleBuffer, kSamples);
+    if (m_frameCount < framesToRead) {
+        qWarning() << "Failed to read chunk samples:"
+                << "actual =" << m_frameCount
+                << ", expected =" << framesToRead;
+        // Adjust the max. readable frame index for future
+        // read requests to avoid repeated invalid reads.
+        *pMaxReadableFrameIndex = frameIndex + m_frameCount;
+    }
+
+    return m_frameCount;
+}
+
+void CachingReaderChunkWorker::copySamples(
+        CSAMPLE* sampleBuffer, SINT sampleOffset, SINT sampleCount) const {
+    DEBUG_ASSERT(0 <= sampleOffset);
+    DEBUG_ASSERT(0 <= sampleCount);
+    DEBUG_ASSERT((sampleOffset + sampleCount) <= frames2samples(m_frameCount));
+    SampleUtil::copy(sampleBuffer, m_sampleBuffer + sampleOffset, sampleCount);
+}
+
+CachingReaderChunk::CachingReaderChunk(
+        CSAMPLE* sampleBuffer)
+        : CachingReaderChunkWorker(sampleBuffer),
+          m_state(FREE),
+          m_pPrev(nullptr),
+          m_pNext(nullptr) {
+}
+
+CachingReaderChunk::~CachingReaderChunk() {
+}
+
+void CachingReaderChunk::init(SINT index) {
+    DEBUG_ASSERT(READ_PENDING != m_state);
+    CachingReaderChunkWorker::init(index);
+    m_state = READY;
+}
+
+void CachingReaderChunk::free() {
+    DEBUG_ASSERT(READ_PENDING != m_state);
+    CachingReaderChunkWorker::init(kDefaultIndex);
+    m_state = FREE;
+}
+
+void CachingReaderChunk::insertIntoListBefore(
+        CachingReaderChunk* pBefore) {
+    DEBUG_ASSERT(nullptr == m_pNext);
+    DEBUG_ASSERT(nullptr == m_pPrev);
+
+    m_pNext = pBefore;
+    if (pBefore) {
+        if (pBefore->m_pPrev) {
+            m_pPrev = pBefore->m_pPrev;
+            DEBUG_ASSERT(m_pPrev->m_pNext == pBefore);
+            m_pPrev->m_pNext = this;
+        }
+        pBefore->m_pPrev = this;
+    }
```

Comment:
Nope, because we insert the new element before pBefore and the tail element 
will remain the same. Only the head may change if pBefore is currently the 
head of the queue, but only the caller knows this.

I think about how to unify these two operations.

On 07/27/2015 01:55 PM, Daniel Schürmann wrote:
>
> In src/cachingreaderchunk.cpp 
> <https://github.com/mixxxdj/mixxx/pull/652#discussion_r35527945>:
>
> > +}
> > +
> > +void CachingReaderChunk::insertIntoListBefore(
> > +        CachingReaderChunk* pBefore) {
> > +    DEBUG_ASSERT(nullptr == m_pNext);
> > +    DEBUG_ASSERT(nullptr == m_pPrev);
> > +
> > +    m_pNext = pBefore;
> > +    if (pBefore) {
> > +        if (pBefore->m_pPrev) {
> > +            m_pPrev = pBefore->m_pPrev;
> > +            DEBUG_ASSERT(m_pPrev->m_pNext == pBefore);
> > +            m_pPrev->m_pNext = this;
> > +        }
> > +        pBefore->m_pPrev = this;
> > +    }
>
> Do we need to adjust the ppTail here?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/652/files#r35527945>.
>